### PR TITLE
Transform OR LIKE chain to multiMatchAny

### LINKED ISSUE #5895: Transform OR LIKE chain to multiSearchAny or multiMatchAny.
**Use case**
`Params LIKE '%testid_145997%' OR Params LIKE '%testid_146118%' OR Params LIKE '%testid_146119%'`

### PR DESCRIPTION
**Pull request by danlark1**
Closes #5895

Changelog category (leave one):
- Performance Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Transform OR LIKE chain to multiMatchAny. Will enable once we have more confidence it works

> Information about CI checks: https://clickhouse.tech/docs/en/development/continuous-integration/

### CODE REVIEW

File: src/Core/Settings.h [hunk 1]
```diff
@@ -436,6 +436,7 @@ class IColumn;
     M(Bool, optimize_normalize_count_variants, true, "Rewrite aggregate functions that semantically equals to count() as count().", 0) \
     M(Bool, optimize_injective_functions_inside_uniq, true, "Delete injective functions of one argument inside uniq*() functions.", 0) \
     M(Bool, convert_query_to_cnf, false, "Convert SELECT query to CNF", 0) \
+    M(Bool, optimize_or_like_chain, false, "Optimize multiple OR LIKE into multiMatchAny", 0) \
```
Comment (kitaisreal) [inline src/Core/Settings.h:439]: Need to enable this setting value to `true`, for single CI run to verify tests. Will change it back to `false` before merge.

**Comment by kitaisreal** [comment]
@Mergifyio update.

**Comment by amosbird** [comment]
Will it disable index analysis?

**Comment by alexey-milovidov** [comment]
@danlark1 

> Feb 27 21:01:52 /build/build_docker/../src/Interpreters/ConvertFunctionOrLikeVisitor.cpp:20:13: error: 'auto expr_list_fn' can be declared as 'auto *expr_list_fn' [readability-qualified-auto,-warnings-as-errors]
Feb 27 21:01:52         if (auto expr_list_fn = child->as<ASTExpressionList>())
Feb 27 21:01:52             ^~~~~
Feb 27 21:01:52             auto *

**Comment by alexey-milovidov** [comment]
@amosbird Yes, it can disable index analysis if either LIKE 'prefix%' is used or if ngram index is used.
That's why we will not enable this setting by default.

**Review by alexey-milovidov — APPROVED** [review]
The code LGTM.

Before merge:
- check if there is a chance of lost aliases;
- disable this setting by default;
- add settings randomization in CI;

**Comment by alexey-milovidov** [comment]
@amosbird @danlark1 

> 00908_bloom_filter_index

Test showed that index stopped working, as expected.

**Comment by alexey-milovidov** [comment]
Nevertheless, this optimization is a nice option to have and the code is well isolated.
I think we can merge...

**Comment by alexey-milovidov** [comment]
@danlark1 I forgot to ask one question... does this optimization help?

**Pull request merged**